contract_comm/election/election.go

```diff
@@ -0,0 +1,208 @@
+          "type": "address"
+        },
+        {
+          "name": "value",
+          "type": "uint256"
+        },
+        {
+          "name": "lesser",
+          "type": "address"
+        },
+        {
+          "name": "greater",
+          "type": "address"
+        }
+      ],
+      "name": "distributeEpochRewards",
+      "outputs": [],
+      "payable": false,
+      "stateMutability": "nonpayable",
+      "type": "function"
+    },
+		    {
+      "constant": true,
+      "inputs": [
+        {
+          "name": "group",
+          "type": "address"
+        },
+        {
+          "name": "totalEpochRewards",
+          "type": "uint256"
+        }
+      ],
+      "name": "getGroupEpochRewards",
+      "outputs": [
+        {
+          "name": "",
+          "type": "uint256"
+        }
+      ],
+      "payable": false,
+      "stateMutability": "view",
+      "type": "function"
+    }
+]`
+
+var electionABI, _ = abi.JSON(strings.NewReader(electionABIString))
+
+func GetElectedValidators(header *types.Header, state vm.StateDB) ([]common.Address, error) {
+	var newValSet []common.Address
+	// Get the new epoch's validator set
+	_, err := contract_comm.MakeStaticCall(params.ElectionRegistryId, electionABI, "electValidators", []interface{}{}, &newValSet, params.MaxGasForElectValidators, header, state)
+	if err != nil {
+		return nil, err
+	}
+	return newValSet, nil
+}
+
+type voteTotal struct {
+	Group common.Address
+	Value *big.Int
+}
+
+func getTotalVotesForEligibleValidatorGroups(header *types.Header, state vm.StateDB) ([]voteTotal, error) {
+	var groups []common.Address
+	var values []*big.Int
+	_, err := contract_comm.MakeStaticCall(params.ElectionRegistryId, electionABI, "getTotalVotesForEligibleValidatorGroups", []interface{}{}, &[]interface{}{&groups, &values}, params.MaxGasForGetEligibleValidatorGroupsVoteTotals, header, state)
+	if err != nil {
+		log.Error("DistributeEpochRewards, error calling getTotalVotesForEligibleValidatorGroups", "err", err)
+	}
+
+	voteTotals := make([]voteTotal, len(groups))
+	for i, group := range groups {
+		log.Info("Got group vote total", "group", group, "value", values[i])
+		voteTotals[i].Group = group
+		voteTotals[i].Value = values[i]
+	}
+	return voteTotals, err
+}
+
+func getGroupEpochRewards(header *types.Header, state vm.StateDB, group common.Address, totalEpochRewards *big.Int) (*big.Int, error) {
+	var groupEpochRewards *big.Int
+	_, err := contract_comm.MakeStaticCall(params.ElectionRegistryId, electionABI, "getGroupEpochRewards", []interface{}{group, totalEpochRewards}, &groupEpochRewards, params.MaxGasForGetGroupEpochRewards, header, state)
+	if err != nil {
+		return nil, err
+	}
+	return groupEpochRewards, nil
+}
+
+func DistributeEpochRewards(header *types.Header, state vm.StateDB, groups []common.Address, totalEpochRewards *big.Int) (*big.Int, error) {
+	totalRewards := big.NewInt(0)
+	voteTotals, err := getTotalVotesForEligibleValidatorGroups(header, state)
```

Comment:
So we have two moments:
 - beginning of epoch (in reality end of the previous one)
 - end of the epoch

It seems we are querying state about eligible groups at the end, when we should do so at the beginning? is this correct?

Reply:
The way we distribute rewards to voters is by increasing the number of active votes for various groups. When we increase the number of votes for an eligible group, we need to sort that group into the list of all eligible groups. In order to know where to sort that group, we need to know the list of all eligible groups and their vote totals, which is what we're fetching here.